Create SLD to agency name and CyHy ID mapping collection

jsf9k opened the pull request:
This data is already compiled and used in-memory in the BOD 18-01 HTTPS and Trustworthy Email reporting.  The CT work that @felddy is doing in [dhs-ncats/cyhy-ct-logs](https://github.com/dhs-ncats/cyhy-ct-logs) requires this data as well, so we will save it in the `scan` database.

jsf9k committed:
Bumped version from 1.0.10 to 1.0.11

jsf9k committed:
Add a script that creates a SLD/CyHy ID mapping collection

This script maps the SLDs to agency names (both as displayed in the
current-federal data) to CyHy IDs using the agencies.csv file.

jsf9k committed:
Add line to run script for SLD to agency name and CyHy ID

jsf9k committed:
Rename new script

jsf9k committed:
Add code to check if agency name in current-federal matches CyHy ID

There is at least one case (AMTRAK) where an agency is not in the
agency mapping file (agencies.csv) because its name as listed in
current-federal.csv is identical to its CyHy ID.  (See AMTRAK, for
instance.)  In that case is_cyhy_stakeholder needs to be set to true.

This commit adds code to perform the check against the CyHy database
and set is_cyhy_stakeholder accordingly.

jsf9k committed:
Add AMTRAK to agencies.csv and revert last commit

A simpler and more efficient to what was done in the last commit is to
simply add AMTRAK to agencies.csv and revert the changes that query
the CyHy database.

dav3r commented on create_sld_to_agency_name_and_id_mapping.py:
```diff
@@ -0,0 +1,90 @@
+CURRENT_FEDERAL_FILE = SHARED_DATA_DIR + 'artifacts/current-federal_modified.csv'
+
+
+def db_from_config(config_filename):
```
This doesn't need to happen today, but at some point we should import this function from the CyHy library.

jsf9k replied:
:+1:

dav3r commented on create_sld_to_agency_name_and_id_mapping.py:
```diff
@@ -0,0 +1,90 @@
+    # Now add our new results
+    res = db.domains.insert_many(records)
+    if len(res.inserted_ids) != len(records):
+        print(f'Unable to write new SLD records from "{db.name}" '
```
I'd like to see this error message also mention how many records were actually written to the database vs. how many were supposed to be written.

dav3r commented on create_sld_to_agency_name_and_id_mapping.py:
```diff
@@ -0,0 +1,90 @@
+    db.domains.delete_many({})
+
+    # Now add our new results
+    res = db.domains.insert_many(records)
```
As discussed, we would like to see the insert/update process look like this:
* Mark all domain documents as `latest:False`
* [Update with upsert](https://docs.mongodb.com/manual/reference/method/db.collection.updateMany/) all new domain documents with `latest:True`

jsf9k replied:
We ended up doing this using the `scan_date` flag.  That way we don't need to add a `latest` boolean.

dav3r commented on create_sld_to_agency_name_and_id_mapping.py:
```diff
@@ -0,0 +1,90 @@
+    # It would be great to check the return code here, but I don't see
+    # anything of use in DeleteResult:
+    # http://api.mongodb.com/python/current/api/pymongo/results.html#pymongo.results.DeleteResult
+    db.domains.delete_many({})
```
This step will no longer be needed once we use update/upsert as described in the next comment.

dav3r reviewed (approved):
Approved, but please see requested changes and implement at an appropriate time

h-m-f-t commented:
This is so great 🇺🇸

jsf9k committed:
Rework logic that adds data to collection

The previous code, although neater, had the issue that someone could
come along and grab the data between when the old rows were dropped
and the new rows were added.  In this new version we use update_one()
with upsert=True to update/create the rows.  Once that is complete we
delete any still-existing rows whose scan_date value is not the most
current.

Once we move to MongoDB 4 I can use a transaction to to atomically (1)
drop all the rows from the collection and (2) use insert_many() to
insert all the new data.  That will be much cleaner!

jsf9k commented:
@dav3r, do you want to take another look since we changed a lot of lines?  If you're still good I'll merge.

I did test this in my `jsf9k` workspace.

dav3r reviewed (approved):
Looking solid now @jsf9k - I agree that transactions are the way to go here. 🏎
